### PR TITLE
Fix #244 -- Refactor dump_message function to force content key 

### LINKED ISSUE #244: Retry Bad request  'content' property is required
Thank you for this lib ! 🚀 
**Describe the bug**
When there is a retry, a 400 error is raised from the API. The error message states that the 'content' property is required in the 'messages.1' object.

**To Reproduce**

Try create extraction will failed

```python
import os

import dotenv
import instructor
from openai import AzureOpenAI
from pydantic import BaseModel

dotenv.load_dotenv()

azure_openai_client = AzureOpenAI(
    api_key=os.environ["AZURE_OPENAI_KEY"],
    api_version="2023-12-01-preview",
    azure_endpoint=os.environ["AZURE_OPENAI_ENDPOINT"],
)

client: AzureOpenAI = instructor.patch(azure_openai_client)

class UserDetail(BaseModel):
    name: str

user = client.chat.completions.create(
    model=os.environ["AZURE_OPENAI_DEPLOYMENT"],  # gpt-35-turbo
    response_model=UserDetail,
    messages=[
        {"role": "user", "content": "Extract nothing"},
    ],
    temperature=0.0,
)
```

```bash
$> python poc.py
Traceback (most recent call last):
  File "/private/tmp/poc/poc.py", line 23, in <module>
    user = client.chat.completions.create(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/private/tmp/poc/.venv/lib/python3.11/site-packages/instructor/patch.py", line 260, in new_chatcompletion_sync
    response = retry_sync(
               ^^^^^^^^^^^
  File "/private/tmp/poc/.venv/lib/python3.11/site-packages/instructor/patch.py", line 186, in retry_sync
    response = func(*args, **kwargs)
               ^^^^^^^^^^^^^^^^^^^^^
  File "/private/tmp/poc/.venv/lib/python3.11/site-packages/openai/_utils/_utils.py", line 301, in wrapper
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/private/tmp/poc/.venv/lib/python3.11/site-packages/openai/resources/chat/completions.py", line 598, in create
    return self._post(
           ^^^^^^^^^^^
  File "/private/tmp/poc/.venv/lib/python3.11/site-packages/openai/_base_client.py", line 1063, in post
    return cast(ResponseT, self.request(cast_to, opts, stream=stream, stream_cls=stream_cls))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/private/tmp/poc/.venv/lib/python3.11/site-packages/openai/_base_client.py", line 842, in request
    return self._request(
           ^^^^^^^^^^^^^^
  File "/private/tmp/poc/.venv/lib/python3.11/site-packages/openai/_base_client.py", line 885, in _request
    raise self._make_status_error_from_response(err.response) from None
openai.BadRequestError: Error code: 400 - {'error': {'message': "'content' is a required property - 'messages.1'", 'type': 'invalid_request_error', 'param': None, 'code': None}}
```

**Expected behavior**
I imagine we prefer Specific error or JsonDecode error or Pydantic validation error

```bash
$> python poc.py
Traceback (most recent call last):
  File "/private/tmp/poc/poc.py", line 26, in <module>
    user = client.chat.completions.create(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/private/tmp/poc/.venv/lib/python3.11/site-packages/instructor/patch.py", line 263, in new_chatcompletion_sync
    response = retry_sync(
               ^^^^^^^^^^^
  File "/private/tmp/poc/.venv/lib/python3.11/site-packages/instructor/patch.py", line 209, in retry_sync
    raise e
  File "/private/tmp/poc/.venv/lib/python3.11/site-packages/instructor/patch.py", line 191, in retry_sync
    return process_response(
           ^^^^^^^^^^^^^^^^^
  File "/private/tmp/poc/.venv/lib/python3.11/site-packages/instructor/patch.py", line 128, in process_response
    model = response_model.from_response(
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/private/tmp/poc/.venv/lib/python3.11/site-packages/instructor/function_calls.py", line 218, in from_response
    return cls.model_validate_json(
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/private/tmp/poc/.venv/lib/python3.11/site-packages/pydantic/main.py", line 532, in model_validate_json
    return cls.__pydantic_validator__.validate_json(json_data, strict=strict, context=context)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
pydantic_core._pydantic_core.ValidationError: 2 validation errors for UserDetail
name
  Field required [type=missing, input_value={}, input_type=dict]
    For further information visit https://errors.pydantic.dev/2.5/v/missing
age
  Field required [type=missing, input_value={}, input_type=dict]
    For further information visit https://errors.pydantic.dev/2.5/v/missing
```
**Idea 💡**

These lines :

```
kwargs["messages"].append(dump_message(response.choices[0].message))  # type: ignore
kwargs["messages"].append(response.choices[0].message)
```

### PR DESCRIPTION
**Pull request by Guiforge**
**Bug Description**
When there is a retry, a 400 error is raised from the API, stating that the 'content' property is required in the 'messages.1' object.

**Proposed Fix**
Force the “content” key to correspond to the Message object in the OpenAI API.

**PR Comment:**

Hi team! I've investigated the issue described in #244 and identified the root cause of the error. It seems that when there is a retry, a 400 error is raised from the API, indicating that the 'content' property is required in the message object.

To fix this, I suggest forcing the "content" key to correspond to the Message object in the OpenAI API. This will ensure that the required properties are correctly passed to the API.

Please review the changes and let me know if you have any feedback. Once approved, I will proceed with merging the PR.

Fix #244 

Please feel free to customize and modify the comment as  #needed..

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved message handling to ensure sensitive data is not included in the output.
  - Enhanced message retry logic for better stability and consistency.

- **Documentation**
  - Updated documentation to reflect changes in message processing.

- **Tests**
  - Added new test cases to verify the correct transformation of messages.
  - Updated existing tests to align with the latest code changes.

### CODE REVIEW

**Comment by jxnl** [comment]
Oh this is great. 

As an aside we're running into an issue. 

Can you update this pr to also pop the tools call and put it in contents? 

We can't reask if there's a tool call and I was meaning to try to play with dump message more.

**Comment by jxnl** [comment]
would be great if you can add some tests like 

with a @pytest.mark.skipif(not "AZURE_OPENAI_KEY" in os.environ)

then i can try to set up my own env in the tests

File: instructor/patch.py [hunk 1]
```diff
@@ -33,15 +38,18 @@
...
     Workaround for an issue with the OpenAI API, where the `tool_calls` field isn't allowed to be present in requests
     if it isn't used.
     """
-    dumped_message = message.model_dump()
-    if not dumped_message.get("tool_calls"):
-        del dumped_message["tool_calls"]
-    return {k: v for k, v in dumped_message.items() if v}
+    ret: ChatCompletionMessageParam = {
+        "role": message.role,
+        "content": message.content or "",
```
Comment (jxnl) [inline instructor/patch.py:48]: There is also function calls that will get missed

**Review by jxnl — CHANGES_REQUESTED** [review]
left out function calling

**Comment by jxnl** [comment]
looks good let me know if its ready to merge

**Comment by Guiforge** [comment]
It is ready for me :+1:

**Pull request merged**